File: sdk/node/examples/advanced_features_demo.ts
#!/usr/bin/env npx tsx
/**
 * A3S Code — Advanced Features Demo (Real LLM)
 *
 * Demonstrates features NOT covered by agentic_loop_demo.js:
 * 1. Direct tool execution (no LLM)
 * 2. Hooks (lifecycle event interception)
 * 3. Queue & lanes (priority-based scheduling)
 * 4. Security provider (taint tracking)
 * 5. Auto-compaction & resilience
 * 6. Persistent memory
 *
 * Usage:
 *     cd crates/code/sdk/node
 *     npx tsx examples/advanced_features_demo.ts
 */

import {
  Agent,
  Session,
  SessionOptions,
  SessionQueueConfig,
  AgentEvent,
  AgentResult,
  ToolResult,
  builtinSkills,
} from '../index.js';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

// ============================================================================
// AdvancedFeaturesDemo
// ============================================================================

class AdvancedFeaturesDemo {
  private readonly agent: Agent;
  private readonly configPath: string;

  constructor(agent: Agent, configPath: string) {
    this.agent = agent;
    this.configPath = configPath;
  }

  // --------------------------------------------------------------------------
  // Static helpers
  // --------------------------------------------------------------------------

  static findConfigPath(): string {
    if (process.env.A3S_CONFIG) return process.env.A3S_CONFIG;
    const homeConfig: string = path.join(os.homedir(), '.a3s', 'config.hcl');
    if (fs.existsSync(homeConfig)) return homeConfig;
    const projectConfig: string = path.join(
      __dirname, '..', '..', '..', '..', '..', '..', '.a3s', 'config.hcl'
    );
    if (fs.existsSync(projectConfig)) return projectConfig;
    throw new Error('Config not found. Create ~/.a3s/config.hcl or set A3S_CONFIG');
  }

  private static separator(title: string): void {
    console.log(`\n${'='.repeat(72)}`);
    console.log(`  ${title}`);
    console.log(`${'='.repeat(72)}\n`);
  }

  private static truncate(text: string, max: number = 200): string {
    text = text.trim();
    return text.length <= max ? text : `${text.substring(0, max)}...`;
  }

  private static makeTempDir(): string {
    return fs.mkdtempSync(path.join(os.tmpdir(), 'a3s-adv-'));
  }

  private static cleanupDir(dir: string): void {
    try { fs.rmSync(dir, { recursive: true, force: true }); } catch (_) {}
  }

  // --------------------------------------------------------------------------
  // Demo 1: Direct Tool Execution
  // --------------------------------------------------------------------------

  async demo1DirectTools(): Promise<void> {
    AdvancedFeaturesDemo.separator('Demo 1: Direct Tool Execution (No LLM)');

    const workspace: string = AdvancedFeaturesDemo.makeTempDir();
    try {
      const session: Session = this.agent.session(workspace, { permissive: true });
      console.log(`  Workspace: ${workspace}\n`);

      // Write
      console.log('  1. Write a file');
      let r: ToolResult = await session.tool('write', { path: 'hello.txt', content: 'Hello from Node.js!\nLine 2.' });
      console.log(`     exit=${r.exitCode} output=${AdvancedFeaturesDemo.truncate(r.output, 60)}`);

      // Read
      console.log('  2. Read the file');
      const content: string = await session.readFile('hello.txt');
      console.log(`     content: ${AdvancedFeaturesDemo.truncate(content, 60)}`);

      // Glob
      console.log('  3. Glob for *.txt');
      const files: string[] = await session.glob('*.txt');
      console.log(`     matches: ${JSON.stringify(files)}`);

      // Bash
      console.log('  4. Run bash command');
      const bashOut: string = await session.bash('wc -l hello.txt');
      console.log(`     output: ${bashOut.trim()}`);

      // Grep
      console.log('  5. Grep for "Node"');
      const grepOut: string = await session.grep('Node');
      console.log(`     output: ${AdvancedFeaturesDemo.truncate(grepOut, 60)}`);

      // Edit
      console.log('  6. Edit the file');
      r = await session.tool('edit', {
        path: 'hello.txt',
        old_string: 'Line 2.',
        new_string: 'Line 2 — edited from Node.js!'
      });
      console.log(`     exit=${r.exitCode}`);

      // Verify
      const final_: string = await session.readFile('hello.txt');
      console.log(`\n  Final content: ${AdvancedFeaturesDemo.truncate(final_, 80)}`);
    } finally {
      AdvancedFeaturesDemo.cleanupDir(workspace);
    }
  }

  // --------------------------------------------------------------------------
  // Demo 2: Hooks
  // --------------------------------------------------------------------------

  async demo2Hooks(): Promise<void> {
    AdvancedFeaturesDemo.separator('Demo 2: Hooks (Lifecycle Event Interception)');

    const workspace: string = AdvancedFeaturesDemo.makeTempDir();
    try {
      const session: Session = this.agent.session(workspace, { permissive: true });

      // Register hooks
      session.registerHook('audit_tools', 'pre_tool_use', null, { priority: 10 });
      session.registerHook('audit_bash', 'post_tool_use', { tool: 'bash' }, { priority: 20 });
      session.registerHook('log_gen', 'generate_start');
      session.registerHook('log_err', 'on_error');

      console.log(`  Registered ${session.hookCount()} hooks`);
      console.log('  * audit_tools  (PreToolUse, priority=10)');
      console.log('  * audit_bash   (PostToolUse, bash only)');
      console.log('  * log_gen      (GenerateStart)');
      console.log('  * log_err      (OnError)\n');

      for await (const event of await session.stream(
        "Create a file test.sh with `echo hello` and run it with bash."
      )) {
        switch (event.type) {
          case 'tool_start':
            console.log(`  [hook] PreToolUse -> ${event.toolName}`);
            break;
          case 'tool_end':
            if (event.toolName === 'bash') {
              console.log(`  [hook] PostToolUse(bash) -> exit=${event.exitCode}`);
            }
            break;
          case 'end':
            console.log(`\n  Done (${event.totalTokens} tokens)`);
            break;
          case 'error':
            console.log(`  Error: ${event.error}`);
            break;
        }
        if (event.type === 'end' || event.type === 'error') break;
      }

      const removed: boolean = session.unregisterHook('audit_tools');
      console.log(`\n  Unregistered audit_tools: ${removed}`);
      console.log(`  Remaining hooks: ${session.hookCount()}`);
    } finally {
      AdvancedFeaturesDemo.cleanupDir(workspace);
    }
  }

  // --------------------------------------------------------------------------
  // Demo 3: Queue & Lanes
  // --------------------------------------------------------------------------

  async demo3QueueLanes(): Promise<void> {
    AdvancedFeaturesDemo.separator('Demo 3: Queue & Lanes (Priority-Based Scheduling)');

    const workspace: string = AdvancedFeaturesDemo.makeTempDir();
    try {
      for (let i = 1; i <= 3; i++) {
        fs.writeFileSync(path.join(workspace, `file${i}.txt`), `Content of file ${i}\n`);
      }

      const qc: SessionQueueConfig = new SessionQueueConfig();
      qc.withLaneFeatures();
      qc.setQueryConcurrency(8);
      qc.setExecuteConcurrency(2);

      const session: Session = this.agent.session(workspace, { permissive: true, queueConfig: qc });

      console.log(`  Queue active: ${session.hasQueue()}`);
      console.log(`  Workspace: ${workspace}\n`);

      const result: AgentResult = await session.send(
        'Read all .txt files and create a combined.md with their contents.'
      );

      console.log(`  Tool calls: ${result.toolCallsCount}`);
      console.log(`  Tokens:     ${result.totalTokens}`);

      const stats: Record<string, unknown> = session.queueStats();
      console.log(`\n  Queue stats: ${JSON.stringify(stats)}`);

      const dlq: Array<Record<string, unknown>> = session.deadLetters();
      console.log(`  Dead letters: ${dlq.length}`);
    } finally {
      AdvancedFeaturesDemo.cleanupDir(workspace);
    }
  }

  // --------------------------------------------------------------------------
  // Demo 4: Security Provider
  // --------------------------------------------------------------------------

  async demo4Security(): Promise<void> {
    AdvancedFeaturesDemo.separator('Demo 4: Security Provider (Taint Tracking)');

    const workspace: string = AdvancedFeaturesDemo.makeTempDir();
    try {
      fs.writeFileSync(
        path.join(workspace, 'secrets.env'),
        'DATABASE_URL=postgres://admin:p4ssw0rd@db.example.com/prod\n' +
        'API_KEY=sk-abc123def456\n' +
        'AWS_SECRET=AKIAIOSFODNN7EXAMPLE\n'
      );

      const session: Session = this.agent.session(workspace, { defaultSecurity: true, permissive: true });

      console.log('  Security: DefaultSecurityProvider enabled');
      console.log('  Features: taint tracking + output sanitization\n');

      const result: AgentResult = await session.send(
        'Read secrets.env and tell me what types of secrets are in it. ' +
        'Do NOT include the actual secret values in your response.'
      );

      console.log(`  Tool calls: ${result.toolCallsCount}`);
      console.log(`  Tokens:     ${result.totalTokens}`);
      console.log(`  Response:   ${AdvancedFeaturesDemo.truncate(result.text, 300)}`);
    } finally {
      AdvancedFeaturesDemo.cleanupDir(workspace);
    }
  }

  // --------------------------------------------------------------------------
  // Demo 5: Auto-Compaction & Resilience
  // --------------------------------------------------------------------------

  async demo5Resilience(): Promise<void> {
    AdvancedFeaturesDemo.separator('Demo 5: Auto-Compaction & Resilience');

    const workspace: string = AdvancedFeaturesDemo.makeTempDir();
    try {
      const session: Session = this.agent.session(workspace, {
        permissive: true,
        maxParseRetries: 3,
        toolTimeoutMs: 30_000,
        circuitBreakerThreshold: 5,
      });

      console.log('  Parse retries:   3');
      console.log('  Tool timeout:    30s');
      console.log('  Circuit breaker: 5\n');

      const result: AgentResult = await session.send(
        'Create a CSV file with 15 rows of sample data (id, name, score), ' +
        'then create a report.md analyzing the data.'
      );

      console.log(`  Tool calls: ${result.toolCallsCount}`);
      console.log(`  Tokens:     ${result.totalTokens}`);

      for (const name of ['data.csv', 'report.md']) {
        const p: string = path.join(workspace, name);
        if (fs.existsSync(p)) {
          console.log(`  ${name} (${fs.statSync(p).size} bytes)`);
        } else {
          console.log(`  ${name} not found`);
        }
      }
    } finally {
      AdvancedFeaturesDemo.cleanupDir(workspace);
    }
  }

  // --------------------------------------------------------------------------
  // Demo 6: Persistent Memory
  // --------------------------------------------------------------------------

  async demo6Memory(): Promise<void> {
    AdvancedFeaturesDemo.separator('Demo 6: Persistent Memory');

    const workspace: string = AdvancedFeaturesDemo.makeTempDir();
    const memoryDir: string = AdvancedFeaturesDemo.makeTempDir();
    try {
      const session: Session = this.agent.session(workspace, { memoryDir, permissive: true });

      console.log(`  Memory dir: ${memoryDir}\n`);

      // Turn 1: teach
      console.log('  [Turn 1] Teaching the agent...');
      const r1: AgentResult = await session.send(
        'Remember this: The project deadline is March 15, 2026. ' +
        'The tech lead is Alice and the PM is Bob.'
      );
      console.log(`    Tokens: ${r1.totalTokens}`);

      // Turn 2: recall
      console.log('  [Turn 2] Asking about remembered info...');
      const r2: AgentResult = await session.send('Who is the tech lead and when is the deadline?');
      console.log(`    Tokens: ${r2.totalTokens}`);
      console.log(`    Answer: ${AdvancedFeaturesDemo.truncate(r2.text, 150)}`);
    } finally {
      AdvancedFeaturesDemo.cleanupDir(workspace);
      AdvancedFeaturesDemo.cleanupDir(memoryDir);
    }
  }

  // --------------------------------------------------------------------------
  // Run All
  // --------------------------------------------------------------------------

  async runAll(): Promise<void> {
    await this.demo1DirectTools();
    await this.demo2Hooks();
    await this.demo3QueueLanes();
    await this.demo4Security();
    await this.demo5Resilience();
    await this.demo6Memory();

    AdvancedFeaturesDemo.separator('All Demos Complete');
  }
}

// ============================================================================
// Main
// ============================================================================

async function main(): Promise<void> {
  console.log('======================================================================');
  console.log('      A3S Code Node.js SDK — Advanced Features Demo (Real LLM)');
  console.log('======================================================================');

  const configPath: string = AdvancedFeaturesDemo.findConfigPath();
  console.log(`\n  Config: ${configPath}`);

  const agent: Agent = await Agent.create(configPath);
  console.log('  Agent:  created\n');

  const demo = new AdvancedFeaturesDemo(agent, configPath);
  await demo.runAll();
}

main().catch((err: unknown) => {
  console.error('Demo failed:', err);
  process.exit(1);
});

File: sdk/python/examples/advanced_features_demo.py
#!/usr/bin/env python3
"""
A3S Code — Model Switching & Direct Tools Demo

Demonstrates:
1. Switching models per session (different providers)
2. Direct tool execution (bypass LLM)
3. Hooks registration and lifecycle events
4. Queue/lane configuration
5. Auto-compaction and resilience settings
6. Security provider

Usage:
    cd crates/code/sdk/python
    python examples/advanced_features_demo.py
"""

import os
import tempfile
from pathlib import Path
from a3s_code import Agent, SessionOptions, SessionQueueConfig, builtin_skills


class AdvancedFeaturesDemo:
    """Demonstrates advanced features of the A3S Code Python SDK."""

    def __init__(self, agent: Agent, config_path: str) -> None:
        self.agent = agent
        self.config_path = config_path

    @staticmethod
    def find_config_path() -> str:
        if env := os.environ.get("A3S_CONFIG"):
            return env
        home_config = Path.home() / ".a3s" / "config.hcl"
        if home_config.exists():
            return str(home_config)
        project_config = (
            Path(__file__).parent.parent.parent.parent.parent.parent
            / ".a3s" / "config.hcl"
        )
        if project_config.exists():
            return str(project_config)
        raise FileNotFoundError("Config not found. Create ~/.a3s/config.hcl or set A3S_CONFIG")

    @staticmethod
    def separator(title: str) -> None:
        print(f"\n{'=' * 72}")
        print(f"  {title}")
        print(f"{'=' * 72}\n")

    @staticmethod
    def truncate(text: str, max_len: int = 200) -> str:
        text = text.strip()
        return text if len(text) <= max_len else f"{text[:max_len]}..."

    # ========================================================================
    # Demo 1: Direct Tool Execution
    # ========================================================================

    def demo_direct_tools(self) -> None:
        self.separator("Demo 1: Direct Tool Execution (No LLM)")

        with tempfile.TemporaryDirectory() as workspace:
            session = self.agent.session(workspace, permissive=True)
            print(f"  Workspace: {workspace}\n")

            # Write a file
            print("  1. Write a file")
            r = session.tool("write", {"path": "hello.txt", "content": "Hello from Python!\nLine 2."})
            print(f"     exit={r.exit_code} output={self.truncate(r.output, 60)}")

            # Read it back
            print("  2. Read the file")
            content = session.read_file("hello.txt")
            print(f"     content: {self.truncate(content, 60)}")

            # Glob
            print("  3. Glob for *.txt")
            files = session.glob("*.txt")
            print(f"     matches: {files}")

            # Bash
            print("  4. Run bash command")
            output = session.bash("wc -l hello.txt")
            print(f"     output: {output.strip()}")

            # Grep
            print("  5. Grep for 'Python'")
            grep_out = session.grep("Python")
            print(f"     output: {self.truncate(grep_out, 60)}")

            # Edit
            print("  6. Edit the file")
            r = session.tool("edit", {
                "path": "hello.txt",
                "old_string": "Line 2.",
                "new_string": "Line 2 -- edited from Python!"
            })
            print(f"     exit={r.exit_code}")

            # Verify
            final_content = session.read_file("hello.txt")
            print(f"\n  Final content: {self.truncate(final_content, 80)}")

    # ========================================================================
    # Demo 2: Hooks
    # ========================================================================

    def demo_hooks(self) -> None:
        self.separator("Demo 2: Hooks (Lifecycle Event Interception)")

        with tempfile.TemporaryDirectory() as workspace:
            session = self.agent.session(workspace, permissive=True)

            # Register hooks
            session.register_hook("audit_tools", "pre_tool_use", config={"priority": 10})
            session.register_hook("audit_bash", "post_tool_use",
                                  matcher={"tool": "bash"}, config={"priority": 20})
            session.register_hook("log_gen", "generate_start")
            session.register_hook("log_err", "on_error")

            print(f"  Registered {session.hook_count()} hooks")
            print("  * audit_tools  (PreToolUse, priority=10)")
            print("  * audit_bash   (PostToolUse, bash only)")
            print("  * log_gen      (GenerateStart)")
            print("  * log_err      (OnError)\n")

            # Execute with hooks active
            for event in session.stream(
                "Create a file test.sh with `echo hello` and run it with bash."
            ):
                if event.event_type == "tool_start":
                    print(f"  [hook] PreToolUse -> {event.tool_name}")
                elif event.event_type == "tool_end":
                    if event.tool_name == "bash":
                        print(f"  [hook] PostToolUse(bash) -> exit={event.exit_code}")
                elif event.event_type == "end":
                    print(f"\n  Done ({event.total_tokens} tokens)")
                    break
                elif event.event_type == "error":
                    print(f"  Error: {event.error}")
                    break

            # Unregister
            removed = session.unregister_hook("audit_tools")
            print(f"\n  Unregistered audit_tools: {removed}")
            print(f"  Remaining hooks: {session.hook_count()}")

    # ========================================================================
    # Demo 3: Queue & Lanes
    # ========================================================================

    def demo_queue_lanes(self) -> None:
        self.separator("Demo 3: Queue & Lanes (Priority-Based Scheduling)")

        with tempfile.TemporaryDirectory() as workspace:
            # Create test files
            for i in range(1, 4):
                Path(workspace, f"file{i}.txt").write_text(f"Content of file {i}\n")

            qc = SessionQueueConfig()
            qc.with_lane_features()
            qc.set_query_concurrency(8)
            qc.set_execute_concurrency(2)

            session = self.agent.session(workspace, permissive=True, queue_config=qc)

            print(f"  Queue active: {session.has_queue()}")
            print(f"  Workspace: {workspace}\n")

            result = session.send(
                "Read all .txt files and create a combined.md with their contents."
            )

            print(f"  Tool calls: {result.tool_calls_count}")
            print(f"  Tokens:     {result.total_tokens}")

            # Queue stats
            stats = session.queue_stats()
            print(f"\n  Queue stats: {stats}")

            # Dead letters
            dlq = session.dead_letters()
            print(f"  Dead letters: {len(dlq)}")

    # ========================================================================
    # Demo 4: Security Provider
    # ========================================================================

    def demo_security(self) -> None:
        self.separator("Demo 4: Security Provider (Taint Tracking)")

        with tempfile.TemporaryDirectory() as workspace:
            # Create a file with sensitive data
            Path(workspace, "secrets.env").write_text(
                'DATABASE_URL=postgres://admin:p4ssw0rd@db.example.com/prod\n'
                'API_KEY=sk-abc123def456\n'
                'AWS_SECRET=AKIAIOSFODNN7EXAMPLE\n'
            )

            opts = SessionOptions()
            opts.default_security = True
            session = self.agent.session(workspace, options=opts, permissive=True)

            print("  Security: DefaultSecurityProvider enabled")
            print("  Features: taint tracking + output sanitization\n")

            result = session.send(
                "Read secrets.env and tell me what types of secrets are in it. "
                "Do NOT include the actual secret values in your response."
            )

            print(f"  Tool calls: {result.tool_calls_count}")
            print(f"  Tokens:     {result.total_tokens}")
            print(f"  Response:   {self.truncate(result.text, 300)}")

    # ========================================================================
    # Demo 5: Auto-Compaction & Resilience
    # ========================================================================

    def demo_resilience(self) -> None:
        self.separator("Demo 5: Auto-Compaction & Resilience")

        with tempfile.TemporaryDirectory() as workspace:
            opts = SessionOptions()
            opts.auto_compact = True
            opts.auto_compact_threshold = 0.7

            session = self.agent.session(
                workspace,
                options=opts,
                permissive=True,
                max_parse_retries=3,
                tool_timeout_ms=30_000,
                circuit_breaker_threshold=5,
            )

            print("  Auto-compact:    enabled (70% threshold)")
            print("  Parse retries:   3")
            print("  Tool timeout:    30s")
            print("  Circuit breaker: 5\n")

            result = session.send(
                "Create a CSV file with 15 rows of sample data (id, name, score), "
                "then create a report.md analyzing the data."
            )

            print(f"  Tool calls: {result.tool_calls_count}")
            print(f"  Tokens:     {result.total_tokens}")

            for name in ["data.csv", "report.md"]:
                p = Path(workspace) / name
                if p.exists():
                    print(f"  {name} ({p.stat().st_size} bytes)")
                else:
                    print(f"  {name} not found")

    # ========================================================================
    # Demo 6: Memory (Persistent)
    # ========================================================================

    def demo_memory(self) -> None:
        self.separator("Demo 6: Persistent Memory")

        with tempfile.TemporaryDirectory() as workspace:
            with tempfile.TemporaryDirectory() as memory_dir:
                opts = SessionOptions()
                opts.memory_dir = memory_dir

                session = self.agent.session(workspace, options=opts, permissive=True)

                print(f"  Memory dir: {memory_dir}\n")

                # Turn 1: teach the agent something
                print("  [Turn 1] Teaching the agent...")
                r1 = session.send(
                    "Remember this: The project deadline is March 15, 2026. "
                    "The tech lead is Alice and the PM is Bob."
                )
                print(f"    Tokens: {r1.total_tokens}")

                # Turn 2: ask about it (memory recall)
                print("  [Turn 2] Asking about remembered info...")
                r2 = session.send("Who is the tech lead and when is the deadline?")
                print(f"    Tokens: {r2.total_tokens}")
                print(f"    Answer: {self.truncate(r2.text, 150)}")

    # ========================================================================
    # Run All Demos
    # ========================================================================

    def run_all(self) -> None:
        print("+" + "=" * 70 + "+")
        print("|     A3S Code Python SDK -- Advanced Features Demo (Real LLM)        |")
        print("+" + "=" * 70 + "+")

        print(f"\n  Config: {self.config_path}")
        print("  Agent:  created\n")

        self.demo_direct_tools()
        self.demo_hooks()
        self.demo_queue_lanes()
        self.demo_security()
        self.demo_resilience()
        self.demo_memory()

        self.separator("All Demos Complete")


if __name__ == "__main__":
    config_path = AdvancedFeaturesDemo.find_config_path()
    agent = Agent.create(config_path)
    demo = AdvancedFeaturesDemo(agent, config_path)
    demo.run_all()
